E0121 overlapping spans / bad placeholder

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
// check-pass: this used to be a stack overflow because of recursion in `usefulness.rs`

macro_rules! long_tuple_arg {
    ([$($t:tt)*]#$($h:tt)*) => {
        long_tuple_arg!{[$($t)*$($t)*]$($h)*}
    };
    ([$([$t:tt $y:tt])*]) => {
        pub fn _f(($($t,)*): ($($t,)*)) {}
    }
}

long_tuple_arg!{[[_ u8]]#}

fn main() {}
```

### Minimal

```rust
macro_rules! foo {
    ($ty:ty) => {
        fn foo(_: $ty, _: $ty) {}
    }
}

foo!(_);

fn main() {}
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
33688d24673db07df084483cbcf0b6bc569ea33a
```

### Error output

```
error[E0121]: the placeholder `_` is not allowed within types on item signatures for functions
  --> 5A2D2B75C2BF81E687B5271EF31D0BB4F73D27EFD91EEA26ABC563EAC0B84418.rs:12:19
   |
12 | long_tuple_arg!{[[_ u8]]#}
   |                   ^
   |                   |
   |                   not allowed in type signatures
   |                   not allowed in type signatures
   |
help: use type parameters instead
   |
8  ~         pub fn _f<T>(($($t,)*): ($($t,)*)) {}
9  |     }
10 | }
11 |
12 ~ long_tuple_arg!{[[TT u8]]#}
   |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0121`.
```

--- a/compiler/rustc_errors/src/diagnostic.rs
+++ b/compiler/rustc_errors/src/diagnostic.rs
@@ -622,16 +622,17 @@ impl Diagnostic {
     pub fn multipart_suggestion_with_style(
         &mut self,
         msg: impl Into<SubdiagnosticMessage>,
-        suggestion: Vec<(Span, String)>,
+        mut suggestion: Vec<(Span, String)>,
         applicability: Applicability,
         style: SuggestionStyle,
     ) -> &mut Self {
-        let mut parts = suggestion
+        suggestion.sort_unstable();
+        suggestion.dedup();
+
+        let parts = suggestion
             .into_iter()
             .map(|(span, snippet)| SubstitutionPart { snippet, span })
             .collect::<Vec<_>>();
-
-        parts.sort_unstable_by_key(|part| part.span);
 
         assert!(!parts.is_empty());
         debug_assert_eq!(

--- a/tests/ui/macros/issue-118048.rs
+++ b/tests/ui/macros/issue-118048.rs
@@ -1,0 +1,10 @@
+macro_rules! foo {
+    ($ty:ty) => {
+        fn foo(_: $ty, _: $ty) {}
+    }
+}
+
+foo!(_);
+//~^ ERROR the placeholder `_` is not allowed within types on item signatures for functions
+
+fn main() {}

--- a/tests/ui/macros/issue-118048.stderr
+++ b/tests/ui/macros/issue-118048.stderr
@@ -1,0 +1,21 @@
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for functions
+  --> $DIR/issue-118048.rs:7:6
+   |
+LL | foo!(_);
+   |      ^
+   |      |
+   |      not allowed in type signatures
+   |      not allowed in type signatures
+   |
+help: use type parameters instead
+   |
+LL ~         fn foo<T>(_: $ty, _: $ty) {}
+LL |     }
+LL | }
+LL | 
+LL ~ foo!(T);
+   |
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0121`.

